dbms/src/Flash/Mpp/GRPCSendQueue.h

```diff
@@ -0,0 +1,264 @@
+    /// Return EMPTY if there is no data in queue and `new_tag` is saved.
+    /// When the next push/finish is called, the `new_tag` will be pushed
+    /// into grpc completion queue.
+    /// Note that any data in `new_tag` mustn't be touched if this function
+    /// returns EMPTY because this `new_tag` may be popped out in another
+    /// grpc thread immediately. By the way, if this completion queue is only
+    /// tied to one grpc thread, this data race will not happen.
+    GRPCSendQueueRes pop(T & data, void * new_tag)
+    {
+        RUNTIME_ASSERT(new_tag != nullptr, log, "new_tag is nullptr");
+
+        auto res = send_queue.tryPop(data);
+        switch (res)
+        {
+        case MPMCQueueResult::OK:
+            return GRPCSendQueueRes::OK;
+        case MPMCQueueResult::FINISHED:
+            return GRPCSendQueueRes::FINISHED;
+        case MPMCQueueResult::EMPTY:
+            // Handle this case later.
+            break;
+        default:
+            RUNTIME_ASSERT(false, log, "Result {} is invalid", res);
+        }
+
+        std::unique_lock lock(mu);
+
+        RUNTIME_ASSERT(status == Status::NONE, log, "status {} is not none", status);
+
+        // Double check if this queue is empty.
+        res = send_queue.tryPop(data);
+        switch (res)
+        {
+        case MPMCQueueResult::OK:
+            return GRPCSendQueueRes::OK;
+        case MPMCQueueResult::FINISHED:
+            return GRPCSendQueueRes::FINISHED;
+        case MPMCQueueResult::EMPTY:
+        {
+            // If empty, change status to WAITING.
+            status = Status::WAITING;
+            tag = new_tag;
+            return GRPCSendQueueRes::EMPTY;
+        }
+        default:
+            RUNTIME_ASSERT(false, log, "Result {} is invalid", res);
+        }
+    }
+
+    /// Finish the queue and kick the grpc completion queue.
+    ///
+    /// For return value meanings, see `MPMCQueue::finish`.
+    bool finish()
+    {
+        auto ret = send_queue.finish();
+        if (ret)
+        {
+            kickCompletionQueue();
+        }
+        return ret;
+    }
+
+    /// Finish and drain the queue.
+    void finishAndDrain()
+    {
+        send_queue.finishAndDrain();
+        kickCompletionQueue();
+    }
+
+private:
+    friend class tests::TestGRPCSendQueue;
+
+    void * kickTagAction()
+    {
+        std::unique_lock lock(mu);
+
+        RUNTIME_ASSERT(status == Status::QUEUING, log, "status {} is not queuing", status);
+        status = Status::NONE;
+
+        return std::exchange(tag, nullptr);
+    }
+
+    /// Wake up its completion queue.
+    void kickCompletionQueue()
+    {
+        {
+            std::unique_lock lock(mu);
+            if (status != Status::WAITING)
+            {
+                return;
+            }
+            RUNTIME_ASSERT(tag != nullptr, log, "status is waiting but tag is nullptr");
+            status = Status::QUEUING;
+        }
+
+        grpc_call_error error = kick_func(&kick_tag);
```

Comment:
is it safe to call `kick_func(&kick_tag)` concurrently?

Reply:
Call `grpc_call_start_batch` with a empty batch is thread-safe. 
PS: I don't know why the comments here https://github.com/grpc/grpc/pull/16362/files says it's thread-compatible but https://github.com/tikv/grpc-rs/pull/217/files#diff-1b4f412f47df1edd4ad45800cfcc9134256c302b000d84646a75e0e7f022a944R765-R768 here says it's thread-safe.

Reply:
Do you mean kicking a NULL tag is safe? how about add a UT for it.